Add HP DC model quirks to AD1884A HDA codec

Add headphone and lineout sensing for the HP Compaq DC7900 ultraslim
desktop, muting the internal mono speaker when appropriate.

Signed-off-by: Claudio Matsuoka <claudio@mandriva.com>
---
 sound/pci/hda/patch_analog.c |   81 +++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 81 insertions(+)

--- linux-2.6.38.3-rc1-alsa/sound/pci/hda/patch_analog.c.orig	2011-04-12 19:52:09.843949402 +0300
+++ linux-2.6.38.3-rc1-alsa/sound/pci/hda/patch_analog.c	2011-04-12 20:20:31.790095804 +0300
@@ -4551,6 +4551,79 @@ static int ad1984a_touchsmart_init(struc
 
 
 /*
+ * HP DC7900
+ * 0x11 - headphone port
+ * 0x12 - line out
+ * 0x13 - internal mono speaker
+ */
+
+/* additional verbs for HPDC model */
+static struct hda_verb ad1884a_hpdc_verbs[] = {
+	/* Port-A (HP) pin - always unmuted */
+	{0x11, AC_VERB_SET_AMP_GAIN_MUTE, AMP_OUT_UNMUTE},
+	/* unsolicited event for pin-sense */
+	{0x11, AC_VERB_SET_UNSOLICITED_ENABLE, AC_USRSP_EN | AD1884A_HP_EVENT},
+	{0x12, AC_VERB_SET_UNSOLICITED_ENABLE, AC_USRSP_EN | AD1884A_HP_EVENT},
+	{0x14, AC_VERB_SET_UNSOLICITED_ENABLE, AC_USRSP_EN | AD1884A_MIC_EVENT},
+	{ } /* end */
+};
+
+/* mute speaker and line out if HP is plugged, or speaker if line out is plugged */
+static void ad1884a_hpdc_automute(struct hda_codec *codec)
+{
+	unsigned int hp_present, spk_present;
+
+	hp_present = snd_hda_codec_read(codec, 0x11, 0,
+				     AC_VERB_GET_PIN_SENSE, 0) & 0x80000000;
+	spk_present = snd_hda_codec_read(codec, 0x12, 0,
+				     AC_VERB_GET_PIN_SENSE, 0) & 0x80000000;
+
+	snd_hda_codec_amp_stereo(codec, 0x12, HDA_OUTPUT, 0,
+				 HDA_AMP_MUTE, hp_present ? HDA_AMP_MUTE : 0);
+	snd_hda_codec_write(codec, 0x12, 0, AC_VERB_SET_EAPD_BTLENABLE,
+			    hp_present ? 0x00 : 0x02);
+
+	snd_hda_codec_amp_stereo(codec, 0x13, HDA_OUTPUT, 0,
+				 HDA_AMP_MUTE, hp_present || spk_present ?
+				 HDA_AMP_MUTE : 0);
+	snd_hda_codec_write(codec, 0x13, 0, AC_VERB_SET_EAPD_BTLENABLE,
+			    hp_present || spk_present ? 0x00 : 0x02);
+}
+
+/* switch to external mic if plugged */
+static void ad1884a_hpdc_automic(struct hda_codec *codec)
+{
+	unsigned int present;
+
+	present = snd_hda_codec_read(codec, 0x14, 0,
+				     AC_VERB_GET_PIN_SENSE, 0) & 0x80000000;
+	snd_hda_codec_write(codec, 0x0c, 0, AC_VERB_SET_CONNECT_SEL,
+			    present ? 0 : 1);
+}
+
+/* unsolicited event for HP DC HP jack sensing */
+static void ad1884a_hpdc_unsol_event(struct hda_codec *codec, unsigned int res)
+{
+	switch (res >> 26) {
+	case AD1884A_HP_EVENT:
+		ad1884a_hpdc_automute(codec);
+		break;
+	case AD1884A_MIC_EVENT:
+		ad1884a_hpdc_automic(codec);
+		break;
+	}
+}
+
+/* initialize HP DC jack-sensing */
+static int ad1884a_hpdc_init(struct hda_codec *codec)
+{
+	ad198x_init(codec);
+	ad1884a_hpdc_automute(codec);
+	ad1884a_hpdc_automic(codec);
+	return 0;
+}
+
+/*
  */
 
 enum {
@@ -4558,6 +4631,7 @@ enum {
 	AD1884A_LAPTOP,
 	AD1884A_MOBILE,
 	AD1884A_THINKPAD,
+	AD1884A_HPDC,
 	AD1984A_TOUCHSMART,
 	AD1984A_PRECISION,
 	AD1884A_MODELS
@@ -4568,6 +4642,7 @@ static const char * const ad1884a_models
 	[AD1884A_LAPTOP]	= "laptop",
 	[AD1884A_MOBILE]	= "mobile",
 	[AD1884A_THINKPAD]	= "thinkpad",
+	[AD1884A_HPDC]		= "hpdc",
 	[AD1984A_TOUCHSMART]	= "touchsmart",
 	[AD1984A_PRECISION]	= "precision",
 };
@@ -4583,6 +4658,7 @@ static struct snd_pci_quirk ad1884a_cfg_
 	SND_PCI_QUIRK_MASK(0x103c, 0xff00, 0x3600, "HP laptop", AD1884A_LAPTOP),
 	SND_PCI_QUIRK_MASK(0x103c, 0xfff0, 0x7010, "HP laptop", AD1884A_MOBILE),
 	SND_PCI_QUIRK(0x17aa, 0x20ac, "Thinkpad X300", AD1884A_THINKPAD),
+	SND_PCI_QUIRK(0x103c, 0x3036, "HP Compaq DC7900", AD1884A_HPDC),
 	SND_PCI_QUIRK(0x103c, 0x2a82, "Touchsmart", AD1984A_TOUCHSMART),
 	{}
 };
@@ -4675,6 +4751,11 @@ static int patch_ad1884a(struct hda_code
 		codec->patch_ops.unsol_event = ad1984a_precision_unsol_event;
 		codec->patch_ops.init = ad1984a_precision_init;
 		break;
+	case AD1884A_HPDC:
+		spec->init_verbs[spec->num_init_verbs++] = ad1884a_hpdc_verbs;
+		codec->patch_ops.unsol_event = ad1884a_hpdc_unsol_event;
+		codec->patch_ops.init = ad1884a_hpdc_init;
+		break;
 	case AD1984A_TOUCHSMART:
 		spec->mixers[0] = ad1984a_touchsmart_mixers;
 		spec->init_verbs[0] = ad1984a_touchsmart_verbs;
